### PR TITLE
Use Brush classes from CoreFX

### PR DESCRIPTION
**Pull request by qmfrederik**
Where possible, use source code from CoreFX for the various Brushes.

I did not move all brushes. For example Mono does additional validation in the `LinearGradientBrush`, which is not done in CoreFX. Some of Mono's test would fail when consuming LinearGradientBrush from CoreFX. We can take care of that later.

Different from the previous PRs, this PR includes code from CoreFx that P/Invokes into gdiplus.

The main work in this PR is to 
* Include the P/Invoke layer (`NativeMethods`, `SafeNativeMethods`, ...) from CoreFX in Mono's System.Drawing.
* Account for different names for some internal properties (e.g. `NativeBrush` vs `NativeObject`). If CoreFX/Mono code is going to co-exist as we migrate to the CoreFX code, we may need this for a while.
* Add some compatibility shims (`SR` and friends)

### CODE REVIEW

**Comment by qmfrederik** [comment]
/cc @akoeplinger @mellinoe

I think this approach will enable us to ingest CoreFX code piece by piece. All feedback is welcome.

**Comment by mellinoe** [comment]
Hey @qmfrederik, thanks for the continued help! Transitioning code piece-by-piece seems like a decent approach to me, and this change looks pretty reasonable overall.

Instead of doing the work here, though, I would suggest that we try to do this over in corefx, on the .NET Core version, once I've merged in https://github.com/dotnet/corefx/pull/20887 (should be very soon). I think it will be a little easier this way:

* We can do everything in one repository, and don't have to worry about updating submodules, etc.
* We will be able to very easily compare to the "baseline Windows version" that is checked into corefx. I think the ultimate goal is to have all of the implementations match as closely as possible to that one.
* We can run automated tests against the Windows version, and we are in the process of writing tests to validate behavior compatibility. In corefx we should also be able to run our tests against the original .NET Framework code on Windows.

I think once the corefx version is in a good state, then we can bring over the changes wholesale, by updating the submodule, referencing the new code, and also referencing the new tests.

@akoeplinger Thoughts?

**Comment by qmfrederik** [comment]
@mellinoe I currently depend on Mono's version of System.Drawing so that's why you'll find me naturally inclined to work in the Mono repository :).

One of the reasons I'm doing the work in the Mono repository is that this one has test coverage in place, whereas corefx only has limited coverage. The tests help highlight some of the implementation differences, fast. For example, I don't think we would have caught the difference in some of the brushes if the work was done in corefx.

Changing the code over in corefx without running Mono's unit test as part of a CI process is bound to create a lot of friction when that work will be integrated in Mono.

In the end, I think that if we do the cleanup in the Mono repository first, dotnet/corefx#20887 would be reduced to a dozen files or so, instead of the 149 files that it consists of now.

**Comment by mellinoe** [comment]
@qmfrederik That is fair enough. I'd like to just do whatever is easiest for everyone and involves the least amount of jumping back and forth between repos. Your point about tests is good, but we are also working on consolidating those over in corefx as well, and we have been trying to augment and clean them as we go. We have the benefit over there of being able to test against the baseline implementation from Windows, as well as the original .NET Framework version. We could port over the tests into corefx before we merge the implementations there, but I can see how that is extra work on top of what is needed if you work in mono.

**Comment by qmfrederik** [comment]
Sure. Most of the work (until this PR) was replacing entire files, I think that can be ported easily to corefx.

Do you think we can get Mono's tests bulk imported in CoreFX (porting to xUnit should be straightforward) and have the refactoring (theory's, naming, style,...) done at a later time? That would make me more comfortable in corefx.

What about we try to do some groundwork in Mono and consolidate in CoreFX once CI is up with libgdiplus and most of Mono's tests are ported over, too?

In the end, the code changes are limited, it's figuring out an approach and the right set of changes that takes up time :)

**Comment by mellinoe** [comment]
> Do you think we can get Mono's tests bulk imported in CoreFX (porting to xUnit should be straightforward) and have the refactoring (theory's, naming, style,...) done at a later time? That would make me more comfortable in corefx.

I think that is a reasonable approach. I would rather have everything in one place, and all efforts focused there, even if it meant that we were merging in code that didn't 100% meet our style and test guidelines yet. We've had similar situations where we ported over tests from our .NET Framework codebase, and later cleaned up and reformatted it.

I'm also fine with continuing to do the consolidation here until we move more of the code into corefx. I can hold off on merging my PR until a little bit more consolidation is done, and then rebase on top of that. Full disclaimer: I'm on vacation for the next week, so I won't have much time to work on it anyways 😄

File: mcs/class/System.Drawing/System.Drawing/SR.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
+//
```
Comment (marek-safar) [inline mcs/class/System.Drawing/System.Drawing/SR.cs]: We generate this file, look how it's done in e.g. mcs/class/System.Core/Makefile

Comment (qmfrederik) [inline mcs/class/System.Drawing/System.Drawing/SR.cs]: Thanks! Should be fixed now.

**Comment by qmfrederik** [comment]
Looks like we'll need to update the API snapshot as well, but I'm not sure what the process is for that:
https://jenkins.mono-project.com/job/test-mono-pull-request-amd64/7267//Public_API_Diff/

**Pull request merged**